core/src/main/java/org/elasticsearch/search/aggregations/metrics/scripted/ScriptedMetricAggregator.java

```diff
@@ -188,6 +222,97 @@ private static Script deepCopyScript(Script script, SearchContext context) {
             return clone;
         }
 
+        @Override
+        protected XContentBuilder internalXContent(XContentBuilder builder, Params builderParams) throws IOException {
+            builder.startObject();
+            if (initScript != null) {
+                builder.field(ScriptedMetricParser.INIT_SCRIPT_FIELD.getPreferredName(), initScript);
+            }
+
+            if (mapScript != null) {
+                builder.field(ScriptedMetricParser.MAP_SCRIPT_FIELD.getPreferredName(), mapScript);
+            }
+
+            if (combineScript != null) {
+                builder.field(ScriptedMetricParser.COMBINE_SCRIPT_FIELD.getPreferredName(), combineScript);
+            }
+
+            if (reduceScript != null) {
+                builder.field(ScriptedMetricParser.REDUCE_SCRIPT_FIELD.getPreferredName(), reduceScript);
+            }
+            if (params != null) {
+                builder.field(ScriptedMetricParser.PARAMS_FIELD.getPreferredName());
+                builder.map(params);
+            }
+            builder.endObject();
+            return builder;
+        }
+
+        @Override
+        protected AggregatorFactory doReadFrom(String name, StreamInput in) throws IOException {
+            Factory factory = new Factory(name);
+            if (in.readBoolean()) {
+                factory.initScript = Script.readScript(in);
+            }
+            if (in.readBoolean()) {
+                factory.mapScript = Script.readScript(in);
+            }
+            if (in.readBoolean()) {
+                factory.combineScript = Script.readScript(in);
+            }
+            if (in.readBoolean()) {
+                factory.reduceScript = Script.readScript(in);
+            }
+            if (in.readBoolean()) {
+                factory.params = in.readMap();
+            }
+            return factory;
+        }
+
+        @Override
+        protected void doWriteTo(StreamOutput out) throws IOException {
+            boolean hasInitScript = initScript != null;
+            out.writeBoolean(hasInitScript);
+            if (hasInitScript) {
+                initScript.writeTo(out);
+            }
+            boolean hasMapScript = mapScript != null;
+            out.writeBoolean(hasMapScript);
+            if (hasMapScript) {
+                mapScript.writeTo(out);
+            }
+            boolean hasCombineScript = combineScript != null;
+            out.writeBoolean(hasCombineScript);
+            if (hasCombineScript) {
+                combineScript.writeTo(out);
+            }
+            boolean hasReduceScript = reduceScript != null;
+            out.writeBoolean(hasReduceScript);
+            if (hasReduceScript) {
+                reduceScript.writeTo(out);
+            }
+            boolean hasParams = params != null;
+            out.writeBoolean(hasParams);
+            if (hasParams) {
+                out.writeMap(params);
+            }
+        }
```

Comment:
should we use writeOptionalStreamable/readOptionalStreamable for serialization instead?